src/env.cc

```diff
@@ -1051,6 +1052,25 @@ void Environment::AddArrayBufferAllocatorToKeepAliveUntilIsolateDispose(
   keep_alive_allocators_->insert(allocator);
 }
 
+bool Environment::RunWeakRefCleanup() {
+  isolate()->ClearKeptObjects();
+
+  TryCatchScope try_catch(this);
+  while (!cleanup_finalization_groups_.empty()) {
+    Local<FinalizationGroup> fg =
+        cleanup_finalization_groups_.front().Get(isolate());
+    cleanup_finalization_groups_.pop();
+    USE(FinalizationGroup::Cleanup(fg));
+    if (try_catch.HasCaught()) {
+      if (!try_catch.HasTerminated()) {
+        try_catch.ReThrow();
```

Comment:
Is this HasCaught() -> HasTerminated() -> ReThrow() logic necessary? Can't you omit the `TryCatchScope` and check `FinalizationGroup::Cleanup(fg).IsNothing()`? That means there's an exception pending.

Reply:
usually we have to rethrow to deal with pending vs scheduled exceptions. I will admit I don't know if it is needed here. I can try to experiment a bit more.